polyfill/index.d.ts

```diff
@@ -0,0 +1,385 @@
+     * absolute time to return:
+     * - `'earlier'`: The earlier time of two possible times
+     * - `'later'`: The later of two possible times
+     * - `'reject'`: Throw a RangeError instead.
+     *
+     * The default is `'earlier'`.
+     *
+     * Compatibility Note: the legacy `Date` object (and libraries like
+     * moment.js and Luxon) gives the same result as `earlier` when turning the
+     * clock back, and `later` when setting the clock forward.
+     *
+     * */
+    disambiguation: 'earlier' | 'later' | 'reject';
+    // TODO: should there be a `compatible` (to legacy `Date`, moment, etc.)
+    //       disambiguation option? See #614.
+  };
+
+  /**
+   * Options for arithmetic operations like `plus()` and `minus()`
+   * */
+  export type ArithmeticOptions = {
+    /**
+     * Controls handling of out-of-range arithmetic results.
+     *
+     * If a result is out of range, then `'constrain'` will clamp the result to
+     * the allowed range, while `'reject'` will throw a RangeError.
+     *
+     * The default is `'constrain'`.
+     */
+    disambiguation: 'constrain' | 'reject';
+    // TODO: may change name from `disambiguation` to `overflow`. See #607.
+  };
+
+  /**
+   * Options to control `Duration.prototype.minus()` behavior
+   * */
+  export type DurationMinusOptions = {
+    /**
+     * Controls how to deal with subtractions that result in negative overflows
+     *
+     * In `'balanceConstrain'` mode, negative fields are balanced with the next
+     * highest field so that none of the fields are negative in the result. If
+     * this is not possible, a `RangeError` is thrown.
+     *
+     * In `'balance'` mode, all fields are balanced with the next highest field,
+     * no matter if they are negative or not.
+     *
+     * TODO: add an example of each mode.
+     *
+     * The default is `'balanceConstrain'`.
+     *
+     */
+    disambiguation: 'balanceConstrain' | 'balance';
+  };
+
+  export interface DifferenceOptions<T extends string> {
+    /**
+     * The largest unit to allow in the resulting `Temporal.Duration` object.
+     *
+     * Valid values may include `'years'`, `'months'`, `'days'`, `'hours'`,
+     * `'minutes'`, and `'seconds'`, although some types may throw an exception
+     * if a value is used that would produce an invalid result. For example,
+     * `hours` is not accepted by `Date.prototype.difference()`.
+     *
+     * The default depends on the type being used.
+     *
+     * TODO: document which types use which defaults.
+     */
+    largestUnit: T;
+  }
+
+  export type DurationLike = {
+    years?: number;
+    months?: number;
+    days?: number;
+    hours?: number;
+    minutes?: number;
+    seconds?: number;
+    milliseconds?: number;
+    microseconds?: number;
+    nanoseconds?: number;
+  };
+
+  export class Duration extends Required<DurationLike> {
+    static from(
+      item: Temporal.Duration | DurationLike | string | object,
```

Comment:
Yep, `object` looks like a bug here. I think when I originally wrote that declaration I wasn't sure if it worked like TimeZone where objects are coerced to strings as a fall-through case.  Duration and non-TZ types don't work like that, correct?

Reply:
The way `from` works is:
- `Temporal.Duration`: the duration is cloned
- `DurationLike`: a new `Temporal.Duration` is constructed using the property values
- `string`: parsed as an ISO string
- anything else: is converted to a string first

For the record, `{}` and `{centuries: 5}` both work, and both give a duration of 0. But maybe this isn't the TypeScript way? (As I mentioned, I'm not much of a TypeScript programmer)

What's the best practice for expressing liberally-accepting JS APIs in TypeScript? Should the type be `any` (reflecting the reality of what is possible to pass to the function) or `Temporal.Duration | DurationLike | string` (being explicit about what cases the function handles)?

Reply:
> * `DurationLike`: a new `Temporal.Duration` is constructed using the property values
> * `string`: parsed as an ISO string
> * anything else: is converted to a string first

@ptomato - what determines "duration-like"?  If an object has no Duration properties, but it has a `toString()` implementation, will Temporal try to call it?  If this is "no", then I'll remove `object` from the accepted types here because `Duration` or `DurationLike` are the only `object` types that should be allowed. 

> For the record, `{}` and `{centuries: 5}` both work, and both give a duration of 0. But maybe this isn't the TypeScript way? (As I mentioned, I'm not much of a TypeScript programmer)

`{}` seems reasonable for an empty duration and it's currently allowed in the TS types because all fields are optional for the `*Like` interfaces.

But if you do specify an object literal like `{centuries: 5}` then TS will complain, but only for object literals. If you want to opt into this behavior, e.g. to create a custom DurationLike type, that's fine. Users can also use a type cast, which means "Hey TS I know what I'm doing, stop complaining"/ Here's an example of all three cases:
 
![image](https://user-images.githubusercontent.com/277214/82839273-b1da4e80-9e83-11ea-859a-024e39916a5e.png)

BTW, a big reason to do this is typo protection. For example, this is also how we get IDE or compiler errors for `{year: 5}` (no "s").

Hey, will calendars expose custom fields, like Japanese `era` on `Date`?  If yes, then the way we'd probably solve this in TS is to make types generic, e.g. `DateTimeLike<CalendarType = Temporal.IsoCalendar>` or something like that. This would enables calendars to expose extra fields to the calendar types while still retaining type safety, and would let users assert that they're expecting dates from a particular calendar to have particular fields.

> What's the best practice for expressing liberally-accepting JS APIs in TypeScript? Should the type be `any` (reflecting the reality of what is possible to pass to the function) or `Temporal.Duration | DurationLike | string` (being explicit about what cases the function handles)?

The latter. The rule of thumb is that the types should match the documentation, not the implementation. When you build types, it's essentially like documentation encoded into type form--hints from the library about what developers *should* do.  Just because something doesn't throw an exception doesn't mean it should be encouraged in the TS types.

Reply:
>  If an object has no Duration properties, but it has a `toString()` implementation, will Temporal try to call it?

`Temporal.Duration.from({ toString() { return 'P2D'; }}).toString() === 'PT0S'`, so I think the answer is "no."

Reply:
OK I'll remove `object` in those cases.